src/riak_core_job_manager.erl

```diff
@@ -0,0 +1,1310 @@
+%%  <li>{@link riak_core_job_service:cfg_idle_min()}</li>
+%%  <li>{@link riak_core_job_service:cfg_idle_max()}</li>
+%%  <li>{@link riak_core_job_service:cfg_recycle()}</li>
+%% </ul>
+%%
+%% These configuration keys and their defaults are defined as macros in
+%% `riak_core_job.hrl'.
+%%
+%% Dynamic reconfiguration is supported by the {@link reconfigure/0} function.
+%%
+-module(riak_core_job_manager).
+-behaviour(gen_server).
+
+% Public API
+-export([
+    cancel/2,
+    config/0,
+    find/1,
+    reconfigure/0,
+    stats/0,
+    stats/1,
+    submit/1
+]).
+
+% Public Types
+-export_type([
+    cfg_concur_max/0,
+    cfg_hist_max/0,
+    cfg_queue_max/0,
+    cfg_mult/0,
+    cfg_prop/0,
+    config/0,
+    mgr_key/0,
+    stat/0,
+    stat_key/0,
+    stat_val/0
+]).
+
+% Work Callback API
+-export([
+    running_job/1
+]).
+
+% Runner Callback API
+-export([
+    cleanup/1,
+    finished/2,
+    running/1,
+    starting/1
+]).
+
+% Gen_server API
+-export([
+    code_change/3,
+    handle_call/3,
+    handle_cast/2,
+    handle_info/2,
+    init/1,
+    start_link/0,
+    terminate/2
+]).
+
+-include("riak_core_job_internal.hrl").
+
+%% ===================================================================
+%% Internal Types
+%% ===================================================================
+
+%
+% WARNING: This file has entries in dialyzer.ignore-warnings
+%
+% Edits may change matching patterns or necessity - be sure to check/update
+% if you're working in that area.
+%
+% The dialyzer attribute below turns off the warning starting in OTP-18, but
+% check occasionally to see if it still matters - it's an erroneous warning
+% from dialyzer that may be fixed in a newer release, as the types are
+% suitably tagged.
+%
+-dialyzer({no_opaque, c_erase/2}).
+
+-define(StatsDict,  orddict).
+-type stats()   ::  ?orddict_t(stat_key(), stat_val()).
+
+% mgr_key() opaque type
+-record(mgrkey, {
+    mgr     ::  pid(),  % manager process
+    key     ::  rkey()  % reference into the 'run' dictionary
+}).
+
+% JobId/Job pair, always kept together
+-record(jrec, {
+    id      ::  jid(),
+    job     ::  job()
+}).
+
+% Full running job record
+-record(rrec, {
+    rref    ::  rref(),
+    jrec    ::  jrec()
+}).
+
+% RunnerRef/RunnerPid, always kept together
+-record(rref, {
+    ref     ::  rkey(),
+    pid     ::  runner()
+}).
+
+% Job queue of JobId/Job pairs
+-record(jq, {
+    c   = 0             ::  non_neg_integer(),      % count
+    d   = queue:new()   ::  ?queue_t(jrec())        % data
+}).
+
+% 'location' dictionary, JobId => location atom or ref
+% count is not maintained
+-record(ld, {
+    d   = dict:new()    ::  ?dict_t(jid(), jloc())  % data
+}).
+
+% 'running' dictionary, RunnerRef => full record
+-record(rd, {
+    c   = 0             ::  non_neg_integer(),      % count
+    d   = dict:new()    ::  ?dict_t(rkey(), rrec()) % data
+}).
+
+-record(state, {
+    rmax                            ::  pos_integer(),
```

Comment:
`rmax` is the maximum allowable size of the `run` collection.
`qmax` is the maximum allowable size of the `que` collection.
`hmax` is the maximum allowable size of the `hist` collection.